_posts/2018-12-06.Rust-1.31.md

```diff
@@ -0,0 +1,338 @@
+conflicting borrow occurs. This makes these sorts of errors far easier to
+debug.
+
+In Rust 1.31, this feature is exclusive to Rust 2018. We hope to backport it
+to Rust 2015 at a later date.
+
+#### Module system changes
+
+The module system is often one of the hardest things for people new to Rust.
+Everyone has their own things that take time to master, of course, but
+there's a root cause for why it's so confusing to many: while there are
+simple and consistent rules defining the module system, their consequences
+can feel inconsistent, counterintuitive and mysterious.
+
+As such, the 2018 edition of Rust introduces a few new module system
+features, but they end up simplifying the module system, to make it more
+clear as to what is going on.
+
+Here's a brief summary:
+
+* `extern crate` is no longer needed in 99% of circumstances.
+* You can import macros with `use`, rather than a `#[macro_use]` attribute.
+* Absolute paths begin with a crate name, where the keyword `crate` refers to the current crate.
+* A `foo.rs` and `foo/` subdirectory may coexist; `mod.rs` is no longer needed when placing submodules in a subdirectory.
+
+These may seem like arbitrary new rules when put this way, but the mental
+model is now significantly simplified overall.
+
+There's a *lot* of details here, so please read [the edition
+guide](https://doc.rust-lang.org/nightly/edition-guide/rust-2018/module-system/path-clarity.html)
+for full details.
+
+### More lifetime elision rules
+
+Let's talk about a feature that's available in both editions: we've added
+some additional elision rules for `impl` blocks and function definitions.
+Code like this:
+
+```rust
+impl<'a> Reader for BufReader<'a> {
+    // methods go here
+}
+`
+
+can now be written like this:
+
+```rust
+impl Reader for BufReader<'_> {
+    // methods go here
+}
+```
+
+The `'_` lifetime still shows that `BufReader` takes a parameter, but we
+don't need to create a name for it anymore.
+
+Lifetimes are still required to be defined in structs. We're considering
+some options here in the future, but have no concrete plans yet.
+
+### `const fn`
+
+There's several ways to define a function in Rust: a regular function with
+`fn`, an unsafe function with `unsafe fn`, an external function with `extern fn`.
+This release adds a new way to qualify a function: `const fn`. It looks like
+this:
+
+```rust
+const fn foo(x: i32) -> i32 {
+    x + 1
+}
+```
+
+A `const fn` can be called like a regular function, but it can also be used
+in any constant context. When it is, it is evaluated at compile time, rather
+than at run time. As an example:
+
+```rust
+const SIX: i32 = foo(5);
+```
+
+This will execute `foo` at compile time, and set `SIX` to `6`.
+
+`const fn`s cannot do everything that normal `fn`s can do; they must
+have deterministic output. This is important for soundness reasons.
+Currently, `const fn`s can do a minimal subset of operations. Here's
```

Comment:
cc @eddyb @centril

Reply:
Looks accurate / what we said in the tracking issue / reference;
Are you looking for something specific?
cc @oli-obk